Testing and repairs including JSON Schema generation enhancements

wendellpiez opened the pull request:
This PR, when ready, will contain:

- XSpecs unit testing XML converter generator XSLT
- XSpec and XProc testing for schema generation and other processes
- InspectorXSLT developments
- Refactoring submodules

And other improvements.

Apologies for the size of this PR! 

### All Submissions:

- [ ] Have you followed the guidelines in our [Contributing](https://github.com/usnistgov/metaschema-xslt/blob/master/CONTRIBUTING.md) document?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/usnistgov/metaschema-xslt/pulls) for the same update/change?
- [ ] Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\]
- [ ] Do all automated CI/CD checks pass?

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you included examples of how to use your new feature(s)?
- [ ] Have you updated all [website](https://pages.nist.gov/metaschema) and readme documentation affected by the changes you made? Changes to the website can be made in the website/content directory of your branch.

galtm committed:
produce-xml-converter.xsl: XSpec tests and minor XSLT improvements (#87)

* XSpec tests and minor XSLT improvements

Co-authored-by: Wendell Piez <wendell.piez@nist.gov>

wendellpiez committed:
JSON Schema generation improvements (#108)

Refactoring submodules; XProc-based testing; XSpec testing of JSON Schema generation; improvements to JSON Schema generation - addressing #105 also #2007.

wendellpiez committed:
June 2024 Integration (#118)

Integrating branches
Co-authored-by: A.J. Stein <alexander.stein@nist.gov>

wendellpiez commented:
#117 is addressed in the file src/converter-gen/Makefile.

In bash, invoke the XSpecs under `testing` in this folder with `make spec-tests` or less directly by `make test`.

It is also invoked under CI/CD, as shown in results.

Planned next is cleanup, build-out and documentation of all testing. Making this easier for devs can be on this punchlist.

galtm commented:
I drilled down into the XSpec Test Results (https://github.com/usnistgov/metaschema-xslt/pull/119/checks?check_run_id=25960758809) by clicking **Raw output** under `There are 229 tests, see "Raw output" for the full list of tests.` I wanted to see if the tests ran that https://github.com/usnistgov/metaschema-xslt/pull/87#issuecomment-1986521844 says didn't run in that earlier PR. I searched the raw results for a few of the `x:expect` labels and they were there. 👍 

(The raw results appear to alphabetize the labels, btw.)

wendellpiez commented:
Excellent, thanks @galtm. Before this branch is merged, more tests will be placed under the `make` logic, is the current plan....

wendellpiez committed:
June 2024 in-progress merge (#120)

Includes updates to unit testing, Inspector XSLT, JSON Schema (#105) and others

Co-authored-by: A.J. Stein <alexander.stein@nist.gov>

wendellpiez committed:
More testing (#122)

Mostly testing support -
* Adding Makefile in compose phase to run XSpec - presently showing errors!
* Extending and reorganizing testing, including removing temporary artifacts from git
* Adding subpipeline mistakenly omitted from the last PR
* New Makefile for testing/tinydata
* Mending XSpecs testing Metaschema composition (regression protection); also a new tinydata sample.
* Update src/testing/tinydata/tiny-data/terms-of-art.xml (AJ Stein)
Co-authored-by: A.J. Stein <alexander.stein@nist.gov>

---------

Co-authored-by: A.J. Stein <alexander.stein@nist.gov>

wendellpiez commented:
This is ready to shake down. It includes lots of new tests, and refactors submodules.

aj-stein-nist commented on src/compose/testing/1_collected/working_metaschema-COLLECTED.xml:
```diff
@@ -75,15 +75,15 @@
    <METASCHEMA abstract="yes"
       module="oscal-test-m1"
-      _base-uri="file:/C:/Users/wap1/Documents/usnistgov/metaschema/toolchains/xslt-M4/compose/testing/module_metaschema.xml">
+      _base-uri="...">
```
Non-blocking: thanks for cleaning these things up, I appreciate it.

aj-stein-nist commented on src/converter-gen/testing/definition-maps/profile-reduce-map-output.xml:
```diff
@@ -0,0 +1,7009 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Non-blocking: as always, I will recommend deleting intermediate outputs if possible to make the noise on changes as tests evolve cleaner and not confuse novice devs in the repo, but I defer to you as the maintainer on this project. 😄

galtm replied:
Hi, @aj-stein-nist . If I'm interpreting this PR correctly, I introduced this file in #87. The purpose of the file is to provide markup that XSpec can use for unit testing of pieces of this pipeline. Can your concern about novice devs be addressed by code comments in the file or a README in the directory?

aj-stein-nist reviewed (approved):
There some really good improvements here. I did some quick integration testing with OSCAL models and it adds some nice cosmetic improvements. I think there are some other changes that can be made for JSON Schema generation, but that can wait for later issues and PRs filed.

Thanks for this big pile of work.
